### DISCUSSION TITLE
Delay in updating user fields

### DISCUSSION THREAD

**Discussion by kevinnammour** [gladly-team/next-firebase-auth#637] [Q&A] [answered]
When updating the user displayName for instance: 
`
await updateProfile(AuthUser.firebaseUser as User, { displayName: values.displayName });
console.log(AuthUser.displayName);
`
This prints the old displayName, how can I get the new displayName? I can't use AuthUser.firebaseUser?.displayName because firebaseUser is null at first and I am using SSR in my app.

**Comment by kmjennison** [comment, marked as answer]
**Quick answer:** After updating the Firebase user's name, call `AuthUser.getIdToken(true)` to force a token refresh, which will update claims and the `AuthUser.displayName` value.

---

This behavior is currently expected. Right now, the AuthUser value is only updated on the client when the ID token changes (see: [`useFirebaseUser`](https://github.com/gladly-team/next-firebase-auth/blob/f1606100987cb473888f47ae9e1f91da4bdb2801/src/useFirebaseUser.js#L138-L142), [`withAuthUser`](https://github.com/gladly-team/next-firebase-auth/blob/f1606100987cb473888f47ae9e1f91da4bdb2801/src/withAuthUser.js#L95-L109)). Changes to things like display name will not automatically update the AuthUser.

One reason this is good is that it will keep the server-side and client-side AuthUser in sync. On the server side, the `displayName` value relies on the Firebase ID token's claims ([source](https://github.com/gladly-team/next-firebase-auth/blob/f1606100987cb473888f47ae9e1f91da4bdb2801/src/createAuthUser.js#L149)), and changes to claims are not automatically propagated to the client. If the ID token isn't refreshed, it's possible to have a server-rendered AuthUser with the old display name and a client-rendered AuthUser with the new display name.

More information on refreshing claims from [Firebase docs](https://firebase.google.com/docs/auth/admin/custom-claims#propagate_custom_claims_to_the_client):

> After new claims are modified on a user via the Admin SDK, they are propagated to an authenticated user on the client side via the ID token in the following ways:
>
>  * A user signs in or re-authenticates after the custom claims are modified. The ID token issued as a result will contain the latest claims.
>  * An existing user session gets its ID token refreshed after an older token expires.
>  * An ID token is force refreshed by calling `currentUser.getIdToken(true)`.

I recommend that, after updating the Firebase user's name, you call `AuthUser.getIdToken(true)` to force a token refresh, which will update claims and the `AuthUser.displayName` value.

It's possible there's a smoother way to handle this kind of change. Any suggestions welcome!

_Edited to provide more detail._

**Reply by kevinnammour** [reply]
So by calling `currentUser.getIdToken(true)`, `AuthUser.displayName` will be updated? The thing is that `AuthUser.firebaseUser.displayName` is updated but not `AuthUser.displayName` when updating the profile, so there is something wrong with AuthUser not being updated with AuthUser.firebaseUser

**Reply by kmjennison** [reply]
@kevinnammour I updated the answer to provide some more context for why this behaves like it does. Let me know if you have any feedback or questions.
